Incentive accounting reform (#2140)
ethereum/eth2.0-specs: specs/accounting_reform/beacon-chain.md

# Incentive accounting  reforms

## Table of contents

**TODO**: Generate table of contents with doctoc.

## Introduction

This is a proposed incentive accounting simplification which replaces `PendingAttestation` with epoch participation flags.

## Constants

### Participation flags

| Name | Value |
| - | - |
| `TIMELY_HEAD_FLAG` | `0` |
| `TIMELY_SOURCE_FLAG` | `1` |
| `TIMELY_TARGET_FLAG` | `2` |

### Participation rewards

| Name | Value |
| - | - |
| `TIMELY_HEAD_NUMERATOR` | `12` |
| `TIMELY_SOURCE_NUMERATOR` | `12` |
| `TIMELY_TARGET_NUMERATOR` | `32` |
| `REWARD_DENOMINATOR` | `64` |

The reward fractions add up to 7/8, leaving the remaining 1/8 for proposer rewards and other future micro-rewards.

### Misc

| Name | Value |
| - | - |

> | `PARTICIPATION_FLAGS_LENGTH` | `8` |

Reviewer: Why is this 8 and not 3?

Reviewer 2: Yeah, think it should be 3
Unless @vbuterin is padding out for future flags, but I don't think that's very valuable because we'd have to hard-fork to add them anyway

| `FLAGS_AND_NUMERATORS` | `((TIMELY_HEAD_FLAG, TIMELY_HEAD_NUMERATOR), (TIMELY_SOURCE_FLAG, TIMELY_SOURCE_NUMERATOR), (TIMELY_TARGET_FLAG, TIMELY_TARGET_NUMERATOR))` |

## Containers

### Modified containers

#### `BeaconState`

```python
class BeaconState(Container):
    # Versioning
    genesis_time: uint64
    genesis_validators_root: Root
    slot: Slot
    fork: Fork
    # History
    latest_block_header: BeaconBlockHeader
    block_roots: Vector[Root, SLOTS_PER_HISTORICAL_ROOT]
    state_roots: Vector[Root, SLOTS_PER_HISTORICAL_ROOT]
    historical_roots: List[Root, HISTORICAL_ROOTS_LIMIT]
    # Eth1
    eth1_data: Eth1Data
    eth1_data_votes: List[Eth1Data, EPOCHS_PER_ETH1_VOTING_PERIOD * SLOTS_PER_EPOCH]
    eth1_deposit_index: uint64
    # Registry
    validators: List[Validator, VALIDATOR_REGISTRY_LIMIT]
    balances: List[Gwei, VALIDATOR_REGISTRY_LIMIT]
    # Randomness
    randao_mixes: Vector[Bytes32, EPOCHS_PER_HISTORICAL_VECTOR]
    # Slashings
    slashings: Vector[Gwei, EPOCHS_PER_SLASHINGS_VECTOR]  # Per-epoch sums of slashed effective balances
    # Participation

>     previous_epoch_participation: List[Bitvector[PARTICIPATION_FLAGS_LENGTH], VALIDATOR_REGISTRY_LIMIT]

Reviewer: I feel like we should swap the order of the validator and flag indices here (make it `List[Bitlist[VALIDATOR_REGISTRY_LIMIT], PARTICIPATION_FLAGS_LENGTH]`): Currently Bitvectors are padded to full bytes when serializing or Merkleizing, meaning that each 3 bit vector will take up a full byte of space.

Author: The problem I see with doing that it that it would increase the amount of Merkling you need to do to process an attestation by ~3x, as you would have to adjust the branches in three positions instead of one. I think on net it would be a reduction in efficiency (as a Merkle branch going into a committee is in most cases longer than the committee itself).

Reviewer: I see, We could also do the following:
1. Change the SSZ spec to allow for tight packing of Bitlists (Don't think we have used vectors of Bitlists, so this should be fine)
2. Make it a Bitlist[PARTICIPATION_FLAGS_LENGTH * VALIDATOR_REGISTRY_LIMIT] and address manually.

I think 2 would be easy to do and would suggest that.

    committee = get_beacon_committee(state, data.slot, data.index)
    assert len(attestation.aggregation_bits) == len(committee)

    if data.target.epoch == get_current_epoch(state):
        epoch_participation = state.current_epoch_participation
        justified_checkpoint = state.current_justified_checkpoint
    else:
        epoch_participation = state.previous_epoch_participation
        justified_checkpoint = state.previous_justified_checkpoint

    # Matching roots
    is_matching_head = data.beacon_block_root == get_block_root_at_slot(state, data.slot)
    is_matching_source = data.source == justified_checkpoint
    is_matching_target = data.target.root == get_block_root(state, data.target.epoch)
    assert is_matching_source

    # Verify signature
    assert is_valid_indexed_attestation(state, get_indexed_attestation(state, attestation))

    # Participation flags
    participation_flags = []
    if is_matching_head and state.slot <= data.slot + MIN_ATTESTATION_INCLUSION_DELAY:
        participation_flags.append(TIMELY_HEAD_FLAG)
    if is_matching_source and state.slot <= data.slot + integer_squareroot(SLOTS_PER_EPOCH):

>         participation_flags.append(TIMELY_SOURCE_FLAG)

Author: Any reason why we're checking `is_matching_source` here when it's already a condition of the attestation getting processed at all? (see `assert is_matching_source` above).

Reviewer 3: The reason is purely cosmetic (as you noticed). It highlights that head, source and target are treated symmetrically just with different constants `X`, `Y`, `Z`.

```
    if is_matching_head and state.slot <= data.slot + X:
        participation_flags.append(TIMELY_HEAD_FLAG)
    if is_matching_source and state.slot <= data.slot + Y:
        participation_flags.append(TIMELY_SOURCE_FLAG)
    if is_matching_target and state.slot <= data.slot + Z:
        participation_flags.append(TIMELY_TARGET_FLAG)
```

Reviewer 2: I support it. Depending on how a client may choose slice up operations having this explicit makes it clear and easy to reason about

    for index in get_eligible_validator_indices(state):
        base_reward = get_base_reward(state, index)
        if index in unslashed_participating_indices:
            if is_in_inactivity_leak(state):
                # Optimal participatition is fully rewarded to cancel the inactivity penalty
                rewards[index] = base_reward * numerator // REWARD_DENOMINATOR
            else:
                rewards[index] = base_reward * numerator * unslashed_participating_increments
                // (active_increments * REWARD_DENOMINATOR)
        else:
            penalties[index] = base_reward * numerator // REWARD_DENOMINATOR
    return rewards, penalties
```

##### `get_inactivity_penalty_deltas`

*Note*: The function `get_inactivity_penalty_deltas` is modified in the selection of matching target indices and the removal of `BASE_REWARDS_PER_EPOCH`.

```python
def get_inactivity_penalty_deltas(state: BeaconState) -> Tuple[Sequence[Gwei], Sequence[Gwei]]:
    """
    Return inactivity reward/penalty deltas for each validator.
    """
    penalties = [Gwei(0) for _ in range(len(state.validators))]
    if is_in_inactivity_leak(state):

>         reward_numerator_sum = sum(numerator for (_, numerator) in FLAGS_AND_NUMERATORS)

Author: ```suggestion
        reward_numerator_sum = TARGET_NUMERATOR + TIMELY_SOURCE_NUMERATOR + TIMELY_HEAD_NUMERATOR // 2
```

        matching_target_indices = get_unslashed_participating_indices(state, TIMELY_TARGET_FLAG, get_previous_epoch(state))
        for index in get_eligible_validator_indices(state):
            # If validator is performing optimally this cancels all attestation rewards for a neutral balance

>             penalties[index] += get_base_reward(state, index) * reward_numerator_sum // REWARD_DENOMINATOR

Reviewer: Given that you have to attest within one slot to get the head reward, and five epochs for the source reward, this seems even harsher than the previous computation (which was already heavily criticised for penalizing even optimally performing validators). In fact, getting the head right within one epoch is impossible at least 33% of the time, meaning that you are guaranteed to be down ~-1/6 already.

I'd suggest scaling this by 1/2, meaning you can be slightly positive still (which I think we have determined in the past is not a problem).

Author: You mean like in the suggestion I just made directly above?

Reviewer: I even suggested being more lenient and dividing the whole `reward_numerator_sum` by 2.

##### `process_attestation`

```python
def process_attestation(state: BeaconState, attestation: Attestation) -> None:
    data = attestation.data
    assert data.target.epoch in (get_previous_epoch(state), get_current_epoch(state))
    assert data.target.epoch == compute_epoch_at_slot(data.slot)
    assert data.slot + MIN_ATTESTATION_INCLUSION_DELAY <= state.slot <= data.slot + SLOTS_PER_EPOCH
    assert data.index < get_committee_count_per_slot(state, data.target.epoch)

    committee = get_beacon_committee(state, data.slot, data.index)
    assert len(attestation.aggregation_bits) == len(committee)
        
    if data.target.epoch == get_current_epoch(state):
        flags = state.current_epoch_flags
        assert data.source == state.current_justified_checkpoint
    else:
        flags = state.previous_epoch_flags
        assert data.source == state.previous_justified_checkpoint    

    # Verify signature
    assert is_valid_indexed_attestation(state, get_indexed_attestation(state, attestation))
        
    # Process flags
    flags_to_set = []
    if data.target.root == get_block_root(state, data.target.epoch):
        flags_to_set.append(FLAG_TARGET)

>         if data.beacon_block_root == get_block_root_at_slot(state, data.slot) and state.slot - data.slot == 1:

Reviewer 2: Use `MIN_ATTESTATION_INCLUSION_DELAY`

```suggestion
        is_correct_head = data.beacon_block_root == get_block_root_at_slot(state, data.slot)
        is_very_timely = state.slot == data.slot + MIN_ATTESTATION_INCLUSION_DELAY
        if is_correct_head and is_very_timely:
```

            flags_to_set.append(FLAG_HEAD_AND_VERY_TIMELY)

>     if state.slot - data.slot <= integer_squareroot(SLOTS_PER_EPOCH):

Reviewer 2: A little bit safer (looking)
```suggestion
    if state.slot <= data.slot + integer_squareroot(SLOTS_PER_EPOCH):
```

        flags_to_set.append(FLAG_TIMELY)

    # Update participation flags
    for participant in get_attesting_indices(state, data, attestation.aggregation_bits):
        active_position = get_active_validator_indices(state, data.target.epoch).index(participant)
        for flag, numerator in FLAGS_AND_NUMERATORS:
            if flag in flags_to_set:
                # Give proposer reward for new flags
                if not flags[active_position][flag]:
                    proposer_reward = Gwei(
                        get_base_reward(state, participant) * numerator
                        // (REWARD_DENOMINATOR * PROPOSER_REWARD_DENOMINATOR)
                    )

>                 increase_balance(state, get_beacon_proposer_index(state), proposer_reward)

Reviewer 2: ```suggestion
                    increase_balance(state, get_beacon_proposer_index(state), proposer_reward)
```

>             flags[active_position][flag] = True

Reviewer 2: ```suggestion
                flags[active_position][flag] = True
```

```

##### `get_unslashed_participant_indices`

```python

> def get_unslashed_participant_indices(state: BeaconState, flag: uint8, epoch: Epoch) -> Set[ValidatorIndex]:

Reviewer 4: 0x017D05EA6E31A91dC4435FbA7B8F95A069577d7a

        for index in get_eligible_validator_indices(state):
            # If validator is performing optimally this cancels all attestation rewards for a neutral balance
            penalties[index] += Gwei(get_base_reward(state, index) * total_usual_reward_numerator // INACTIVITY_PENALTY_QUOTIENT))
            if index not in matching_target_attesting_indices:
                effective_balance = state.validators[index].effective_balance
                penalties[index] += Gwei(effective_balance * get_finality_delay(state) // INACTIVITY_PENALTY_QUOTIENT)

    # No rewards associated with inactivity penalties
    rewards = [Gwei(0) for _ in range(len(state.validators))]
    return rewards, penalties
```

##### `process_rewards_and_penalties`

```python
def process_rewards_and_penalties(state: BeaconState) -> None:
    # No rewards are applied at the end of `GENESIS_EPOCH` because rewards are for work done in the previous epoch
    if get_current_epoch(state) == GENESIS_EPOCH:
        return

    flag_deltas = [
        get_standard_flag_deltas(state, flag, numerator) for (flag, numerator) in FLAGS_AND_NUMERATORS
    ]
    

>     for (rewards, penalties) in flag_deltas + [get_inactivity_penalty_deltas(state)]:

Reviewer 2: ```suggestion
    all_deltas = flag_deltas + [get_inactivity_penalty_deltas(state)]
    for (rewards, penalties) in all_deltas:
```

    committee = get_beacon_committee(state, data.slot, data.index)
    assert len(attestation.aggregation_bits) == len(committee)
        
    if data.target.epoch == get_current_epoch(state):
        flags = state.current_epoch_flags
        assert data.source == state.current_justified_checkpoint
    else:
        flags = state.previous_epoch_flags
        assert data.source == state.previous_justified_checkpoint    

    # Verify signature
    assert is_valid_indexed_attestation(state, get_indexed_attestation(state, attestation))
        
    # Process flags
    flags_to_set = []
    if data.target.root == get_block_root(state, data.target.epoch):
        flags_to_set.append(FLAG_TARGET)

        is_correct_head = data.beacon_block_root == get_block_root_at_slot(state, data.slot)
        is_very_timely = state.slot == data.slot + MIN_ATTESTATION_INCLUSION_DELAY
        if is_correct_head and is_very_timely:
            flags_to_set.append(FLAG_HEAD_AND_VERY_TIMELY)

    if state.slot <= data.slot + integer_squareroot(SLOTS_PER_EPOCH):
        flags_to_set.append(FLAG_TIMELY)

    # Update participation flags

>     for participant in get_attesting_indices(state, data, attestation.aggregation_bits):
>         active_position = get_active_validator_indices(state, data.target.epoch).index(participant)

Reviewer 5: Take pyspec for example, it seems that `state` object will be changed when setting `flags[active_position][flag] = True` with mutable side effect and the LRU cache keys of `get_active_validator_indices` will be different for each `participant`.

A slight refactoring may make implementers happier:

```suggestion
    active_validator_indices = get_active_validator_indices(state, data.target.epoch)
    for participant in get_attesting_indices(state, data, attestation.aggregation_bits):
        active_position = active_validator_indices.index(participant)
```

        for flag, numerator in FLAGS_AND_NUMERATORS:
            if flag in flags_to_set and not flags[active_position][flag]:
                flags[active_position][flag] = True
                # Give proposer reward for new flags
                proposer_reward = Gwei(
                    get_base_reward(state, participant) * numerator
                    // (REWARD_DENOMINATOR * PROPOSER_REWARD_DENOMINATOR)
                )
                increase_balance(state, get_beacon_proposer_index(state), proposer_reward)
```

##### `get_unslashed_participant_indices`

```python
def get_unslashed_participant_indices(state: BeaconState, flag: uint8, epoch: Epoch) -> Set[ValidatorIndex]:
    assert epoch in [get_current_epoch(state), get_previous_epoch(state)]

>     flags = state.current_epoch_reward_flags if epoch == get_current_epoch(state) else state.previous_epoch_reward_flags

Reviewer 5: ```suggestion
    flags = state.current_epoch_flags if epoch == get_current_epoch(state) else state.previous_epoch_flags
```

        if index in unslashed_participant_indices:
            if is_in_inactivity_leak(state):
                # Since full base reward will be canceled out by inactivity penalty deltas,
                # optimal participation receives full base reward compensation here.
                rewards[index] += get_base_reward(state, index) * numerator * total_balance // REWARD_DENOMINATOR
            else:
                rewards[index] += (
                    (get_base_reward(state, index) * total_participating_balance // total_balance)
                    * numerator // REWARD_DENOMINATOR
                )
        else:
            penalties[index] += get_base_reward(state, index) * numerator // REWARD_DENOMINATOR
    return rewards, penalties
```

##### `get_inactivity_penalty_deltas`

```python
def get_inactivity_penalty_deltas(state: BeaconState) -> Tuple[Sequence[Gwei], Sequence[Gwei]]:
    """
    Return inactivity reward/penalty deltas for each validator.
    Note: function exactly the same as Phase 0 other than the selection of `matching_target_attesting_indices`
    """
    penalties = [Gwei(0) for _ in range(len(state.validators))]
    

>     total_usual_reward_numerator = sum([numerator for (flag, numerator) in FLAGS_AND_NUMERATORS])

Reviewer 5: ```suggestion
    total_usual_reward_numerator = sum(numerator for (_, numerator) in FLAGS_AND_NUMERATORS)
```
